src/kokkos_abstraction.hpp

```diff
@@ -214,476 +296,374 @@ inline void kokkos_dispatch(ParallelScanDispatch, Args &&...args) {
+  static constexpr std::size_t Rank = bound_translator::Rank;
+
+  template <typename ExecSpace>
+  inline void dispatch(std::string name, ExecSpace exec_space, Bounds &&...bounds,
+                       Function function, Args &&...args, const int scratch_level = 0,
+                       const std::size_t scratch_size_in_bytes = 0) {
+    constexpr std::size_t Ninner =
+        dispatch_type::TeamPattern::Nvector + dispatch_type::TeamPattern::Nthread;
+    constexpr auto pattern_tag = LoopPatternTag<dispatch_type::GetPatternTag()>();
+    static_assert(
+        !std::is_same_v<decltype(pattern_tag), LoopPatternTag<LP::undef>> &&
+            !always_false<Tag, Pattern>,
+        "Loop pattern & tag combination not recognized in DispatchType::GetPatternTag");
+
+    constexpr bool isSimdFor =
+        std::is_same_v<LoopPatternSimdFor, base_type<decltype(pattern_tag)>>;
+    auto bound_arr = bound_translator().GetIndexRanges(std::forward<Bounds>(bounds)...);
+    if constexpr (isSimdFor) {
+      static_assert(!isSimdFor || (isSimdFor && std::is_same_v<ExecSpace, HostExecSpace>),
+                    "SimdFor pattern only supported in HostExecSpace");
+      SimdFor(std::make_index_sequence<Rank - 1>(), function, bound_arr);
+    } else {
+      dispatch_impl(pattern_tag, std::make_index_sequence<Rank - Ninner>(),
+                    std::make_index_sequence<Ninner>(), name, exec_space, bound_arr,
+                    function, std::forward<Args>(args)..., scratch_level,
+                    scratch_size_in_bytes);
+    }
+  }
 
-// 4D loop using TeamPolicy loop with inner ThreadVectorRange
-template <typename Tag, typename Function>
-inline void par_dispatch(LoopPatternTPTVR, const std::string &name,
-                         DevExecSpace exec_space, const int nl, const int nu,
-                         const int kl, const int ku, const int jl, const int ju,
-                         const int il, const int iu, const Function &function) {
-  // TODO(pgrete) if exec space is Cuda,throw error
-  const int Nn = nu - nl + 1;
-  const int Nk = ku - kl + 1;
-  const int Nj = ju - jl + 1;
-  const int NkNj = Nk * Nj;
-  const int NnNkNj = Nn * Nk * Nj;
-  Kokkos::parallel_for(
-      name, team_policy(exec_space, NnNkNj, Kokkos::AUTO),
-      KOKKOS_LAMBDA(team_mbr_t team_member) {
-        int n = team_member.league_rank() / NkNj;
-        int k = (team_member.league_rank() - n * NkNj) / Nj;
-        int j = team_member.league_rank() - n * NkNj - k * Nj + jl;
-        n += nl;
-        k += kl;
-        Kokkos::parallel_for(Kokkos::ThreadVectorRange<>(team_member, il, iu + 1),
-                             [&](const int i) { function(n, k, j, i); });
-      });
-}
+  template <std::size_t... Is>
+  using sequence = std::integer_sequence<std::size_t, Is...>;
+
+  template <typename ExecSpace, std::size_t... OuterIs, std::size_t... InnerIs>
+  inline void dispatch_impl(LoopPatternTag<LP::flat>, sequence<OuterIs...>,
```

Comment:
I would add a comment before each of these `dispatch_impl` overloads saying which is which. I know the `LoopPatternTag<...>` is there, but I think that is hard to see if you don't know what you are looking for.